src/openms/source/ANALYSIS/QUANTITATION/AbsoluteQuantitation.cpp

```diff
@@ -0,0 +1,438 @@
+            }
+          }
+          else
+          {
+            quant_methods_it->second.getTransformationModel(transformation_model,transformation_model_params);
+            calculated_concentration = applyCalibration(
+              unknowns[feature_it].getSubordinates()[sub_it],
+              empty_feature,
+              quant_feature_name,transformation_model,transformation_model_params);
+          }
+
+          // add new metadata (calculated_concentration, concentration_units) to the component
+          unknowns[feature_it].getSubordinates()[sub_it].setMetaValue("calculated_concentration",calculated_concentration);
+          quant_methods_it->second.getConcentrationUnits(concentration_units);
+          unknowns[feature_it].getSubordinates()[sub_it].setMetaValue("concentration_units",concentration_units);
+          // calculate the bias?
+        }
+        else 
+        {
+          LOG_INFO << "Component " << component_name << " does not have a quantitation method.";
+          LOG_INFO << "No concentration will be calculated.";
+          unknowns[feature_it].getSubordinates()[sub_it].setMetaValue("calculated_concentration","");
+          unknowns[feature_it].getSubordinates()[sub_it].setMetaValue("concentration_units","");
+        }
+      }
+    }
+    // }
+  }
+
+  void AbsoluteQuantitation::optimizeCalibrationCurveBruteForce(
+    const std::vector<AbsoluteQuantitationStandards::featureConcentration> & component_concentrations,
+    const String & feature_name,
+    const String & transformation_model,
+    const Param & transformation_model_params,
+    Param & optimized_params)
+  {
+    
+    std::vector<double> biases;
+    double r2 = 0.0;
+    bool bias_check;
+
+    //TODO use internal params
+    size_t min_points = 4;
+    double max_bias = 30.0;
+    double min_r2 = 0.9; 
+    size_t max_outliers = 1;  
+
+    size_t n_loops;
+    std::vector<AbsoluteQuantitationStandards::featureConcentration>::const_iterator component_start_it;
+    std::vector<AbsoluteQuantitationStandards::featureConcentration>::const_iterator component_end_it;
+
+    // sort from min to max concentration
+    std::vector<AbsoluteQuantitationStandards::featureConcentration> component_concentrations_sorted = component_concentrations;
+    std::sort(component_concentrations_sorted.begin(), component_concentrations_sorted.end(),
+      [](AbsoluteQuantitationStandards::featureConcentration lhs, AbsoluteQuantitationStandards::featureConcentration rhs)
+      {
+        return lhs.actual_concentration < rhs.actual_concentration;    
+      }
+    );
+
+    // loop from all points to min_points
+    for (size_t n_points = component_concentrations.size(); n_points >= min_points; --n_points)
+    {
+      n_loops = component_concentrations.size() - n_points;
+      for (size_t  component_it = 0; component_it < n_loops; ++component_it)
+      {
+        // TODO: support for outliers
+        // // loop through max_outliers
```

Comment:
@dmccloskey we have some pretty neat support for outlier detection and removal over at the RT alignment code. Please have a look at OpenSwathRTNormalizer and more specifically at OpenMS/ANALYSIS/OPENSWATH/MRMRTNormalizer.h I can work with you to make that code more generic but I think it already is -- e.g. you have a `std::vector<std::pair<double, double> > ` and it will find non-outlier points using either chauvenet, an iterative method based on the residual or jackknife aproach or RANSAC. There is also a publication describing these methods, Parker et al 2015

Reply:
Can you send me the citation?  I am not able to find the Parker et al 2015 study.

This is a good idea.  There are a few criteria that would be good to have (e.g., minimum bias/accuracy for each point in the linear model) that I don't see yet in any of the methods.  However, I can see consolidating such an algorithm in the MRMRTNormalizer class as beneficial for future work as well.

Let me know your thoughts on what you mean by making it more general.

Reply:
This is not a huge concern now, but in the future it might be good to add support for more complex models (i.e., Quadratic and Hill) that are also quite common in the pharmaceutical and clinical realm.  A brute force method would always work, but some of the more specialized methods may not.

Reply:
@dmccloskey this is the study: https://www.ncbi.nlm.nih.gov/pubmed/26199342 my idea was just that this will give you an easy way to perform outlier detection and removal. Also feel free to implement further methods for outlier detection.